x-pack/filebeat/input/httpjson/chain.go

```diff
@@ -56,38 +56,63 @@
 
 package httpjson
 
-import (
-	"github.com/elastic/elastic-agent-libs/mapstr"
-	"github.com/elastic/elastic-agent-libs/transport/httpcommon"
-)
-
 // chainConfig for chain request.
 // Following contains basic call structure for each call after normal httpjson
 // call.
 type chainConfig struct {
-	Step stepConfig `config:"step" validate:"required"`
+	Step  *stepConfig  `config:"step,omitempty"`
+	While *whileConfig `config:"while,omitempty"`
 }
 
 // stepConfig will contain basic properties like, request.url,
 // request.method and replace parameter. Each step: request.url
 // will contain replace string with original URL to make a skeleton for the
 // call request.
 type stepConfig struct {
-	Request  requestChainConfig  `config:"request"`
+	Auth     *authConfig         `config:"auth"`
+	Request  requestConfig       `config:"request" validate:"required"`
 	Response responseChainConfig `config:"response,omitempty"`
 	Replace  string              `config:"replace,omitempty"`
 }
 
-type requestChainConfig struct {
-	URL        *urlConfig       `config:"url" validate:"required"`
-	Method     string           `config:"method" validate:"required"`
-	Body       *mapstr.M        `config:"body"`
-	Transforms transformsConfig `config:"transforms"`
-
-	Transport httpcommon.HTTPTransportSettings `config:",inline"`
+// whileConfig will contain basic properties like auth parameters, request parameters,
+// response parameters , a replace parameter and an expression parameter called till.
+// While is similar to stepConfig with the addition of till. Till holds an expression
```

Comment:
> While is similar to stepConfig with the addition of till.

This is informative. Would it be worthwhile rolling these two types into one and having the repetition behaviour be conditional on `.Till` being non-nil?

Reply:
@efd6 Yes that can be easily done , but the main motivation behind making these separate steps was to keep the overhead low for the future . I fear if we keep adding different attributes to "step" , overtime it might become too complex to maintain and also loose the simplicity of user configuration. Hence went for an approach keeping "separation of concerns" as a guiding principle.

Reply:
Also the name "while" gives meaning to the "till" attribute which plays nicely from a user perspective, and with a recent update , made it such that  "till" is now a mandatory attribute for while config.